k2/csrc/ragged_ops.cu

```diff
@@ -793,4 +799,86 @@ RaggedShape TrivialShape(ContextPtr &c, int32_t num_elems) {
   return RaggedShape2(&row_splits, &row_ids, num_elems);
 }
 
+static Array1<int32_t> GetTransposeReorderingCpu(Ragged<int32_t> &src,
+                                                 int32_t num_cols) {
+  const int32_t *row_splits1_data = src.RowSplits(src.NumAxes() - 1).Data();
+  const int32_t *row_ids1_data = src.RowIds(src.NumAxes() - 1).Data();
+  const int32_t *value_data = src.values.Data();
+  auto lambda_comp = [=](int32_t a_idx01, int32_t b_idx01) -> bool {
+    int32_t a_idx0 = row_ids1_data[a_idx01];
+    int32_t b_idx0 = row_ids1_data[b_idx01];
+
+    int32_t a_col_index = value_data[a_idx01];
+    int32_t b_col_index = value_data[b_idx01];
+
+    if (a_col_index < b_col_index) return true;  // sort by column indexes
+    if (a_col_index > b_col_index) return false;
+
+    // now we have a_col_index == b_col_index
+    if (a_idx0 < b_idx0) return true;  // sort by row indexes
+    if (a_idx0 > b_idx0) return false;
+
+    // now we have a_idx0 == b_idx0 && a_col_index == b_col_index
+    // this entry is duplicated in the sparse matrix.
+    return false;  // we can return either true or false here.
+  };
+  Array1<int32_t> ans(src.Context(), src.values.Dim());
+  int32_t *ans_data = ans.Data();
+  int32_t n = ans.Dim();
+  for (int32_t i = 0; i != n; ++i) ans_data[i] = i;
+
+  std::sort(ans_data, ans_data + n, lambda_comp);
+  return ans;
+}
+
+Array1<int32_t> GetTransposeReordering(Ragged<int32_t> &src, int32_t num_cols) {
+  ContextPtr &context = src.Context();
+  if (src.NumAxes() < 2) {
+    // src is empty
+    return Array1<int32_t>(context, 0);
+  }
+
+  DeviceType device_type = context->GetDeviceType();
+  if (device_type == kCpu) {
+    return GetTransposeReorderingCpu(src, num_cols);
+  }
+
+  K2_CHECK_EQ(device_type, kCuda);
+
+  const int32_t *row_splits1_data = src.RowSplits(src.NumAxes() - 1).Data();
+  const int32_t *row_ids1_data = src.RowIds(src.NumAxes() - 1).Data();
+  const int32_t *value_data = src.values.Data();
+
+  auto lambda_set_value = [] __device__(int32_t idx01) -> int32_t {
+    return idx01;
+  };
+  int32_t n = src.values.Dim();
+  Array1<int32_t> ans(context, n, lambda_set_value);
```

Comment:
With https://github.com/k2-fsa/k2/blob/d4ca5714316774d80be5f7ea713892d087b17ae0/k2/csrc/array_ops.h#L319

You can just do Array1<int32_t> ans = Range(c,n,0);

Reply:
fixed, thanks.